### ISSUE TITLE
 ERROR: EndOfDirectory - called with an invalid handle.

### ISSUE THREAD

**Issue by cancan101** [jbeluch/xbmcswift2#37]
Currently I have a route which calls "plugin.set_resolved_url(path)"

I then either return nothing (None) or an empty list. In either case I see:

ERROR: EndOfDirectory - called with an invalid handle.

in the log. The video DOES correctly start playing, so I am not sure what the effects of this error are.

**Comment by jbeluch** [comment]
Try this:

```
return plugin.set_resolved_url(path)
```

**Comment by jbeluch** [comment]
Let me know if this fixes the problem for you.

**Comment by cancan101** [comment]
Now I get:

23:44:36 T:2869681008   ERROR:  AddItems - called with an invalid handle.
23:44:36 T:2869681008   ERROR:  EndOfDirectory - called with an invalid handle.

**Comment by jbeluch** [comment]
Hmm, is this view being executed via RunPlugin (from a context menu item)?

**Comment by cancan101** [comment]
No.

``` python
play_url = plugin.url_for(endpoint="playLecture",...)
item ={ 'path': play_url, 'is_playable': True, ...}
return [item]

@plugin.route('...')
def playLecture(...):
   path = ...
   ret = plugin.set_resolved_url(path)
   return ret
```

bear in mind the video does play correctly, I just see those ERRORs in log.

**Comment by jbeluch** [comment]
Hmm, can you log the value of plugin.handle:

```
plugin.log.info('Handle: ' + plugin.handle)
```

**Comment by cancan101** [comment]
00:08:21 T:2980027248  NOTICE: [plugin.video.xxx] Handle: 0
00:08:21 T:2980027248   ERROR:  AddItems - called with an invalid handle.
00:08:21 T:2980027248   ERROR:  EndOfDirectory - called with an invalid handle

**Comment by jbeluch** [comment]
Can you send me the complete addon source? Gotta poke around, not sure what's going on.

**Comment by ulion** [comment]
By my old experience, it possible corrected with this:

> looks the is_playable is set True, that's correct.
> but the 'info' dict is missing, that will cause xbmcswift2 not call setInfo(info_type, info) and then the xbmc will not think the item url > need be resolved so it callback the plugin without a new handle, that will cause the error in the log.
> so, at least set 'info': {'title': 'some title'} for the item together with is_playable set.

But I'm wrong, after read xbmc plugin-directory code, found the invalid handle error is caused by two plugin/addon calls inter-mixed, that means before script A call finished, a B call is inited, and then A call finished, the handle vector  is shrinked, then when B call do something with the old handle, it is not existed in the handle vector.
anyway, we need avoid it, and I my previous info works for me but don't known why, it's strange, the xbmc plugin api always was.

**Comment by ulion** [comment]
OK, now I known what happened, here is it:
it's a xbmcswift bug, but just a harmless one.
both endofdirectory and setresolvedurl let xbmc plugindirectory exit the wait loop and delete the handle, but xbmcswift did not prevent the auto call to endofdirectory (which will do nothing because the handle already be deleted) even after a setresolvedurl call.

**Comment by dersphere** [comment]
I had the same "-1" handle issue in a plugin where I call the add-on from the context menu.
To prevent others from hours of debugging here is one possible reason:
There are two different usable context menu calls:

```
def _action(*args, **kwargs):
    return 'XBMC.RunPlugin(%s)' % plugin.url_for(*args, **kwargs)

def _view(*args, **kwargs):
    return 'XBMC.Container.Update(%s)' % plugin.url_for(*args, **kwargs)
```

_action() should be used when the plugin should do some (background) actions and the current view needs to stay.
_view() should be used when the plugin view is going to be replaced.

**Comment by jbeluch** [comment]
Will be fixed in #68. The issue is a result of a bug that ulion describes above.
